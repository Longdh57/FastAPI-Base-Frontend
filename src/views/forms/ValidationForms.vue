<template>
  <CCard>
    <CCardHeader>
      <CIcon name="cil-notes"/> Form Validation
      <a class="badge badge-danger" href="https://coreui.io/pro/vue/">CoreUI Pro</a>
      <div class="card-header-actions">
        <a 
          class="card-header-action" 
          href="https://github.com/vuelidate/vuelidate" 
          target="_blank" 
          rel="noreferrer noopener"
        >
          <small class="text-muted">docs</small>
        </a>
      </div>
    </CCardHeader>
    <CCardBody>
      <CLink 
        href="https://github.com/vuelidate/vuelidate" 
        target="_blank" 
        rel="noreferrer noopener"
      >
        Vuelidate
      </CLink> 
      provides <cite>Simple, lightweight model-based validation for Vue.js. </cite>
      In this view Vuelidate features are integrated with CoreUI Vue form components.
      <hr>
      <CRow>
        <CCol lg="6">
          <!--<h6>Simple Form</h6>-->
          <CForm>
            <CInput
              label="First Name"
              :lazy="false"
              :value.sync="$v.form.firstName.$model"
              :isValid="checkIfValid('firstName')"
              placeholder="First Name"
              autocomplete="given-name"
              invalidFeedback="This is a required field and must be at least 2 characters"
            />

            <CInput
              label="Last Name"
              :lazy="false"
              :value.sync="$v.form.lastName.$model"
              :isValid="checkIfValid('lastName')"
              placeholder="Last Name"
              autocomplete="family-name"
              invalidFeedback="This is a required field and must be at least 1 character"
            />

            <CInput
              label="User Name"
              :lazy="false"
              :value.sync="$v.form.userName.$model"
              :isValid="checkIfValid('userName')"
              placeholder="User Name"
              autocomplete="username"
              invalidFeedback="This is a required field and must be at least 5 character"
            />

            <CInput
              label="Email"
              type="email"
              :lazy="false"
              :value.sync="$v.form.email.$model"
              :isValid="checkIfValid('email')"
              placeholder="Email"
              autocomplete="email"
              invalidFeedback="This is a required field and must be valid e-mail address"
            />

            <CRow>
              <CCol md="6">
                <CInput
                  :isValid="checkIfValid('password')"
                  :value.sync="$v.form.password.$model"
                  label="Password"
                  type="password"
                  placeholder="Password"
                  autocomplete="new-password"
                  invalidFeedback="Required password containing at least: number, uppercase and lowercase letter, 8 characters"
                />
              </CCol>
              <CCol md="6">
                <CInput
                  :isValid="checkIfValid('password')"
                  :value.sync="$v.form.confirmPassword.$model"
                  label="Confirm Password"
                  type="password"
                  placeholder="Password"
                  autocomplete="new-password"
                  invalidFeedback="Passwords must match"
                />
              </CCol>
            </CRow>
            <CInputCheckbox
              :isValid="checkIfValid('accept')"
              :checked.sync="$v.form.accept.$model"
              label="I accept the terms of use"
              invalidFeedback="You must accept before submitting"
              custom
              class="mb-4"
            />
            <CButton 
              color="primary" 
              :disabled="!isValid || submitted"
              @click="submit"
            >
              Submit
            </CButton>
            <CButton 
              class="ml-1"  
              color="success" 
              :disabled="isValid"
              @click="validate"
            >
              Validate
            </CButton>
            <CButton 
              class="ml-1"
              color="danger"
              :disabled="!isDirty"
              @click="reset"
            >
              Reset
            </CButton>
          </CForm>
          <br/>
        </CCol>

        <CCol lg="6">
          <CCard :class="`bg-${submitted ? 'info' : 'secondary' }`">
            <pre>{{formString}}</pre>
          </CCard>
        </CCol>
      </CRow>
    </CCardBody>
  </CCard>
</template>

<script>
import { validationMixin } from "vuelidate"
import { required, minLength, email, sameAs, helpers } from "vuelidate/lib/validators"

export default {
  name: 'ValidationForms',
  data() {
    return {
      form: this.getEmptyForm(),
      submitted: false
    }
  },
  computed: {
    formString () { return JSON.stringify(this.form, null, 4) },
    isValid () { return !this.$v.form.$invalid },
    isDirty () { return this.$v.form.$anyDirty },
  },
  mixins: [validationMixin],
  validations: {
    form: {
      firstName: {
        required,
        minLength: minLength(3)
      },
      lastName: {
        required,
        minLength: minLength(2)
      },
      userName: {
        required,
        minLength: minLength(5)
      },
      email: {
        required,
        email
      },
      password: {
        required,
        minLength: minLength(8),
        strongPass: helpers.regex('strongPass', /(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{8,}/)
      },
      confirmPassword: {
        required,
        sameAsPassword: sameAs("password")
      },
      accept: {
        required,
        mustAccept: val => val
      }
    }
  },
  methods: {
    checkIfValid (fieldName) {
      const field = this.$v.form[fieldName]
      if (!field.$dirty) {
        return null
      } 
      return !(field.$invalid || field.$model === '')
    },

    submit () {
      if (this.isValid) {
        this.submitted = true
      }
    },

    validate () {
      this.$v.$touch()
    },

    reset () {
      this.form = this.getEmptyForm()
      this.submitted = false
      this.$v.$reset()
    },
    
    getEmptyForm () {
      return {
        firstName: "",
        lastName: "",
        userName: "",
        email: "",
        password: "",
        confirmPassword: "",
        accept: false
      }
    }
  }
}
</script>
