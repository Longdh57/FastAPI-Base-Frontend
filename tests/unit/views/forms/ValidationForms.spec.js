import Vue from 'vue'
import { shallowMount } from '@vue/test-utils'
import CoreuiVue from '@coreui/vue-pro'
import ValidationForms from '@/views/forms/ValidationForms'


Vue.use(CoreuiVue)

describe('ValidationForms.vue', () => {
  it('has a name', () => {
    expect(ValidationForms.name).toMatch('ValidationForms')
  })
  it('is Vue instance', () => {
    const wrapper = shallowMount(ValidationForms)
    expect(wrapper.vm).toBeTruthy()
  })
  test('renders correctly', () => {
    const wrapper = shallowMount(ValidationForms)
    expect(wrapper.element).toMatchSnapshot()
  })
})
