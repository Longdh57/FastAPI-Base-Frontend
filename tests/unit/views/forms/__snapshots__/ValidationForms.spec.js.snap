// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidationForms.vue renders correctly 1`] = `
<ccard-stub>
  <ccardheader-stub>
    <cicon-stub
      name="cil-notes"
    />
     Form Validation
    
    <a
      class="badge badge-danger"
      href="https://coreui.io/pro/vue/"
    >
      CoreUI Pro
    </a>
     
    <div
      class="card-header-actions"
    >
      <a
        class="card-header-action"
        href="https://github.com/vuelidate/vuelidate"
        rel="noreferrer noopener"
        target="_blank"
      >
        <small
          class="text-muted"
        >
          docs
        </small>
      </a>
    </div>
  </ccardheader-stub>
   
  <ccardbody-stub>
    <clink-stub
      activeclass="router-link-active"
      event="click"
      exactactiveclass="router-link-exact-active"
      href="https://github.com/vuelidate/vuelidate"
      rel="noreferrer noopener"
      routertag="a"
      target="_blank"
    >
      
      Vuelidate
    
    </clink-stub>
     
    provides 
    <cite>
      Simple, lightweight model-based validation for Vue.js. 
    </cite>
    
    In this view Vuelidate features are integrated with CoreUI Vue form components.
    
    <hr />
     
    <crow-stub
      gutters="true"
      tag="div"
    >
      <ccol-stub
        lg="6"
        tag="div"
      >
        <cform-stub>
          <cinput-stub
            autocomplete="given-name"
            invalidfeedback="This is a required field and must be at least 2 characters"
            label="First Name"
            placeholder="First Name"
            type="text"
            value=""
          />
           
          <cinput-stub
            autocomplete="family-name"
            invalidfeedback="This is a required field and must be at least 1 character"
            label="Last Name"
            placeholder="Last Name"
            type="text"
            value=""
          />
           
          <cinput-stub
            autocomplete="username"
            invalidfeedback="This is a required field and must be at least 5 character"
            label="User Name"
            placeholder="User Name"
            type="text"
            value=""
          />
           
          <cinput-stub
            autocomplete="email"
            invalidfeedback="This is a required field and must be valid e-mail address"
            label="Email"
            placeholder="Email"
            type="email"
            value=""
          />
           
          <crow-stub
            gutters="true"
            tag="div"
          >
            <ccol-stub
              md="6"
              tag="div"
            >
              <cinput-stub
                autocomplete="new-password"
                invalidfeedback="Required password containing at least: number, uppercase and lowercase letter, 8 characters"
                label="Password"
                lazy="400"
                placeholder="Password"
                type="password"
                value=""
              />
            </ccol-stub>
             
            <ccol-stub
              md="6"
              tag="div"
            >
              <cinput-stub
                autocomplete="new-password"
                invalidfeedback="Passwords must match"
                label="Confirm Password"
                lazy="400"
                placeholder="Password"
                type="password"
                value=""
              />
            </ccol-stub>
          </crow-stub>
           
          <cinputcheckbox-stub
            class="mb-4"
            custom="true"
            invalidfeedback="You must accept before submitting"
            label="I accept the terms of use"
          />
           
          <cbutton-stub
            activeclass="router-link-active"
            color="primary"
            disabled="true"
            event="click"
            exactactiveclass="router-link-exact-active"
            routertag="a"
            target="_self"
            type="button"
          >
            
            Submit
          
          </cbutton-stub>
           
          <cbutton-stub
            activeclass="router-link-active"
            class="ml-1"
            color="success"
            event="click"
            exactactiveclass="router-link-exact-active"
            routertag="a"
            target="_self"
            type="button"
          >
            
            Validate
          
          </cbutton-stub>
           
          <cbutton-stub
            activeclass="router-link-active"
            class="ml-1"
            color="danger"
            disabled="true"
            event="click"
            exactactiveclass="router-link-exact-active"
            routertag="a"
            target="_self"
            type="button"
          >
            
            Reset
          
          </cbutton-stub>
        </cform-stub>
         
        <br />
      </ccol-stub>
       
      <ccol-stub
        lg="6"
        tag="div"
      >
        <ccard-stub
          class="bg-secondary"
        >
          <pre>
            {
    "firstName": "",
    "lastName": "",
    "userName": "",
    "email": "",
    "password": "",
    "confirmPassword": "",
    "accept": false
}
          </pre>
        </ccard-stub>
      </ccol-stub>
    </crow-stub>
  </ccardbody-stub>
</ccard-stub>
`;
